### PR TITLE
check ERROR_NO_MORE_FILES to prevent walkDir[Rec] to quit prematurely for windows

### PR DESCRIPTION
**Pull request by oskca**
According to [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364428(v=vs.85).aspx), only `FindNextFile(h, f) ==0` and `GetLastError() == ERROR_NO_MORE_FILES ` truely indicates there is no more file to find.

### CODE REVIEW

File: lib/pure/os.nim [hunk 1]
```diff
@@ -672,7 +672,7 @@ template walkCommon(pattern: string, filter) =
           if dotPos < 0 or idx >= ff.len or ff[idx] == '.' or
               pattern[dotPos+1] == '*':
             yield splitFile(pattern).dir / extractFilename(ff)
-        if findNextFile(res, f) == 0'i32: break
+        if findNextFile(res, f) == 0'i32 and getLastError() == 18: break # ERROR_NO_MORE_FILES=18
```
Comment (dom96) [inline lib/pure/os.nim]: Add this constant to ``winlean``.

Comment (dom96) [reply lib/pure/os.nim]: If it's a different error it should be raised. Use ``raiseOsError()`` for that.

Comment (oskca) [reply lib/pure/os.nim]: if `raiseOsError()`, then the finding process will be stopped and has no change to continue, and `walkDir[Rec]` will never have the chance to walk the whole directory

Comment (dom96) [reply lib/pure/os.nim]: yes, that's what should happen. Errors shouldn't be ignored, unless they can be handled by this code.

File: lib/pure/os.nim [hunk 2]
```diff
@@ -782,7 +785,10 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
             let xx = if relative: extractFilename(getFilename(f))
                      else: dir / extractFilename(getFilename(f))
             yield (k, xx)
-          if findNextFile(h, f) == 0'i32 and getLastError() == 18: break # ERROR_NO_MORE_FILES=18
+          if findNextFile(res, f) == 0'i32:
+            let errCode = getLastError()
+            if errCode == ERROR_NO_MORE_FILES: break
+            else: raiseOSError(errCode, "findNextFile failed")
```
Comment (oskca) [inline lib/pure/os.nim]: raiseOSError os Error for other cases

**Comment by oskca** [comment]
yes, all errors should be raised

**Comment by data-man** [comment]
I think that we need ways to get a list of files/dirs without raising errors.
This is need for writing a file manager, e.g.
I have some ideas for RFC.

**Comment by oskca** [comment]
Looking forward to you RFC @data-man

**Comment by dom96** [comment]
> I think that we need ways to get a list of files/dirs without raising errors.
>This is need for writing a file manager, e.g.
> I have some ideas for RFC.

You could always just catch the error.

But if you're writing a file manager then you should be using the OS api directly anyway. The stdlib doesn't need to handle specialised use cases such as this.

**Comment by dom96** [comment]
There is no need for the string inside raiseOSError so I removed it.

**Comment by dom96** [comment]
Error is unrelated. Merging.

**Comment by data-man** [comment]
@oskca 

> Looking forward to your RFC

#7077

**Pull request merged**